### PR TITLE
Add backward induction and beta =1 #172

### LINKED ISSUE #172: DiscreteDP: Add backward_induction
As in QuantEcon/QuantEcon.py#244

* [x] Let `DiscreteDP` accept `beta=1`
* [x] Add a method `backward_induction` for finite horizon problems

### PR DESCRIPTION
**Pull request by tokupooh**
This pull request is related with #172.

I try the followings.

- add `backward_induction` method and explanation.
- allows  `beta = 1` for DiscreteDP` 
- write test codes for them.

close #172.

### CODE REVIEW

File: src/markov/ddp.jl [hunk 1]
```diff
@@ -501,6 +505,40 @@ function solve(ddp::DiscreteDP{T}, v_init::AbstractVector{T},
...
+- `vs::Array{Float64,2}`: Matrix that contains the optimal value function 
+  at each period
+- `sigmas::Array{Int,2}`: Matrix that contains the optimal policy function
+  at each period
+"""
+function backward_induction(ddp::DiscreteDP{T}, J::Integer,
+                            v_term::AbstractVector{T}=
+                            zeros(num_states(ddp))) where {T}
+    n = num_states(ddp)
+    vs = zeros(J+1,n)
```
Comment (oyamad) [inline src/markov/ddp.jl]: If `T` is `Rational`, the eltype of `vs` should be `Rational` too.

```suggestion
    S = typeof(zero(T)/one(T))
    vs = Matrix{S}(undef, J+1, n)
```

File: src/markov/ddp.jl [hunk 2]
```diff
@@ -501,6 +505,40 @@ function solve(ddp::DiscreteDP{T}, v_init::AbstractVector{T},
...
+    vs[end,:] = v_term
+    v_temp = v_term
+    sigmas = zeros(Int, J, n)
```
Comment (oyamad) [inline src/markov/ddp.jl]:
```suggestion
    sigmas = Matrix{Int}(undef, J, n)
```

File: src/markov/ddp.jl [hunk 3]
```diff
@@ -501,6 +505,40 @@ function solve(ddp::DiscreteDP{T}, v_init::AbstractVector{T},
...
+    sigma_temp = sigmas[end,:]
+    @inbounds for j in J+1: -1: 2
+        v_temp, sigma_temp = bellman_operator!(ddp, v_temp, sigma_temp)
+        vs[j-1,:], sigmas[j-1,:] = v_temp, sigma_temp
+    end
```
Comment (oyamad) [inline src/markov/ddp.jl]: Is it possible to avoid allocating the temporary arrays?

File: src/markov/ddp.jl [hunk 4]
```diff
@@ -501,6 +505,40 @@ function solve(ddp::DiscreteDP{T}, v_init::AbstractVector{T},
...
+
+- `ddp::DiscreteDP{T}` : Object that contains the Model Parameters
+- `J::Integer`: Number of decision periods
+- `v_term::AbstractVector{T}=zeros(num_states(ddp))`: Terminal value function,
+  of length equal to n (the number of states)  
+
+##### Returns
+
+- `vs::Array{Float64,2}`: Matrix that contains the optimal value function 
+  at each period
```
Comment (oyamad) [inline src/markov/ddp.jl]: Describe the shape (size) of `vs` explicitly, as in the Python version.

File: src/markov/ddp.jl [hunk 5]
```diff
@@ -501,6 +505,40 @@ function solve(ddp::DiscreteDP{T}, v_init::AbstractVector{T},
...
+- `sigmas::Array{Int,2}`: Matrix that contains the optimal policy function
+  at each period
```
Comment (oyamad) [inline src/markov/ddp.jl]: Describe the shape (size) of `sigmas` explicitly, as in the Python version.

File: src/markov/ddp.jl (hunk 3 above)
Comment (tokupooh) [reply src/markov/ddp.jl]: We can rewrite this as 
```julia
@inbounds for j in J+1: -1: 2
        vs[j-1, :], sigmas[j-1,:] =  @views bellman_operator!(ddp,vs[j,:],sigmas[j-1,:])
    end
```
Thanks. I'll fix it.

Comment (tokupooh) [reply src/markov/ddp.jl]: Here, we can rewrite this as 
```julia
@inbounds for j in J+1: -1: 2
    @views bellman_operator!(ddp,vs[j,:],vs[j-1,:],sigmas[j-1,:])
end
```

Thanks! I'll fix it.

Comment (oyamad) [reply src/markov/ddp.jl]: By the way, Julia being column-major, the values at period j should be stored the j-th *column* (`vs` and `sigmas` should be of shape `(n, J+1)` and `(n, J)`).

Comment (tokupooh) [reply src/markov/ddp.jl]: For internal consistency, I adopt the above written style, but I follow your suggestion.

**Comment by oyamad** [comment]
Will you also add a test with `Rational` inputs?

(Define the `R` and `Q` arrays with `Rational` first, then run the test with it; in the second test, convert them into `Float64`.)

**Comment by sglyon** [comment]
@oyamad I'd be happy to follow something more official.

I believe @jlperla and his team were working on a style guide. Perhaps they included a discussion of docstrings conventions to follow?

**Comment by tokupooh** [comment]
Thanks.  Fix them.

File: src/markov/ddp.jl [hunk 6]
```diff
@@ -501,6 +507,60 @@ function solve(ddp::DiscreteDP{T}, v_init::AbstractVector{T},
...
+"""
+    backward_induction(ddp, J[, v_term=zeros(num_states(ddp))])
+
+Solve by backward induction a ``J``-period finite horizon discrete dynamic 
+program with stationary reward ``r`` and transition probability functions ``q``
+and discount factor ``\\beta \\in [0, 1]``.
+
+The optimal value functions ``v^{\\ast}_1, \\ldots, v^{\\ast}_{J+1}`` and policy
+functions ``\\sigma^{\\ast}_1, \\ldots, \\sigma^{\\ast}_J`` are obtained by 
+``v^{\\ast}_J = v_J``, and
```
Comment (oyamad) [inline src/markov/ddp.jl]:
```suggestion
``v^{\\ast}_{J+1} = v_{J+1}``, and
```

File: src/markov/ddp.jl [hunk 7]
```diff
@@ -501,6 +507,60 @@ function solve(ddp::DiscreteDP{T}, v_init::AbstractVector{T},
...
+
+```math
+v^{\\ast}_{j-1}(s) = \\max_{a \\in A(s)} r(s, a) +
```
Comment (oyamad) [inline src/markov/ddp.jl]:
```suggestion
v^{\\ast}_j(s) = \\max_{a \\in A(s)} r(s, a) +
```

File: src/markov/ddp.jl [hunk 8]
```diff
@@ -501,6 +507,60 @@ function solve(ddp::DiscreteDP{T}, v_init::AbstractVector{T},
...
+\\beta \\sum_{s' \\in S} q(s'|s, a) v^{\\ast}_j(s')
```
Comment (oyamad) [inline src/markov/ddp.jl]:
```suggestion
\\beta \\sum_{s' \\in S} q(s'|s, a) v^{\\ast}_{j+1}(s')
```

File: src/markov/ddp.jl [hunk 9]
```diff
@@ -501,6 +507,60 @@ function solve(ddp::DiscreteDP{T}, v_init::AbstractVector{T},
...
+\\quad (s \\in S)
+```
+and
+```math 
+\\sigma^{\\ast}_{j-1}(s) \\in \\operatorname*{arg\\,max}_{a \\in A(s)}
```
Comment (oyamad) [inline src/markov/ddp.jl]:
```suggestion
\\sigma^{\\ast}_j(s) \\in \\operatorname*{arg\\,max}_{a \\in A(s)}
```

File: src/markov/ddp.jl [hunk 10]
```diff
@@ -501,6 +507,60 @@ function solve(ddp::DiscreteDP{T}, v_init::AbstractVector{T},
...
+            r(s, a) + \\beta \\sum_{s' \\in S} q(s'|s, a) v^*_j(s')
```
Comment (oyamad) [inline src/markov/ddp.jl]:
```suggestion
            r(s, a) + \\beta \\sum_{s' \\in S} q(s'|s, a) v^*_{j+1}(s')
```

File: src/markov/ddp.jl [hunk 11]
```diff
@@ -501,6 +507,60 @@ function solve(ddp::DiscreteDP{T}, v_init::AbstractVector{T},
...
+            \\quad (s \\in S)
+```
+
+for ``j= J, \\ldots, 1``, where the terminal value function ``v_J`` is
```
Comment (oyamad) [inline src/markov/ddp.jl]:
```suggestion
for ``j= J, \\ldots, 1``, where the terminal value function ``v_{J+1}`` is 
```

File: src/markov/ddp.jl [hunk 12]
```diff
@@ -501,6 +507,60 @@ function solve(ddp::DiscreteDP{T}, v_init::AbstractVector{T},
...
+```
+and
+```math 
+\\sigma^{\\ast}_{j-1}(s) \\in \\operatorname*{arg\\,max}_{a \\in A(s)}
+            r(s, a) + \\beta \\sum_{s' \\in S} q(s'|s, a) v^*_j(s')
+            \\quad (s \\in S)
+```
+
+for ``j= J, \\ldots, 1``, where the terminal value function ``v_J`` is 
+exogenously given.
```
Comment (oyamad) [inline src/markov/ddp.jl]:
```suggestion
exogenously given by `v_term`.
```

**Comment by tokupooh** [comment]
Fix them.

**Comment by oyamad** [comment]
@sglyon Can I merge this PR?

**Comment by sglyon** [comment]
Looks great to me, thank you for following up.

Sorry I forgot to merge this after it was approved

**Pull request merged**